### PR TITLE
GPT2LMScorer batching optimization

### PR DESCRIPTION
**Pull request by dldk-gael**
Following the split of #4, I propose here to optimize the way GPT2LMScorer evaluated multiples sentences by offering the possibility to input the sentences by batches.

I also add several tests specific to this new batching feature in test_gpt2.  

I notice that, it is changing the score value, but really at the margin. For instance : 
```
scorer = GPT2LMScorer("gpt2")
sentences = [
    "I am good",
    "I have a big amount of money.",
    "This is the best day of my life.",
    "Let us finish this quickly!"
]
scorer.batch_size = 1
scores_batch_1 = scorer.sentence_score(sentences, reduce='gmean')
print(scores_batch_1)

scorer.batch_size = 4
scores_batch_4 = scorer.sentence_score(sentences, reduce='gmean')
print(scores_batch_4)
```

Will give : 
```
[0.0023175945971161127, 0.027517564594745636, 0.06669076532125473, 0.00807127170264721]
[0.002317607868462801, 0.02751740626990795, 0.06669124215841293, 0.00807106401771307]
```

### CODE REVIEW

**Comment by dldk-gael** [comment]
Hi Simone, 
Is is still something you are considering to add?

**Comment by simonepri** [comment]
Yes, I'm definitely interested, it's just that I didn't have time yet to review the changes (I'm working on my thesis as well 😬).

**Comment by simonepri** [comment]
@dldk-gael I refactored the in internal API to make this PR easier.
I also added more tests so that we can be more confident that everything works as expected.

If you have time, can you update this PR to make it working with the new API?

I tried to re-implement the batching optimization using the new API and the transformers batch_encoding_plus function, but it's not working yet (it gives the wrong probabilities).
If you want you can try to fix the code below instead.

Thanks!


```python
from typing import *  # pylint: disable=wildcard-import,unused-wildcard-import

import torch
from transformers import AutoTokenizer, AutoConfig, GPT2LMHeadModel
from transformers.tokenization_utils import BatchEncoding

from .abc.transformers import TransformersLMScorer


class GPT2LMScorer(TransformersLMScorer):
    # @overrides
    def _build(self, model_name: str, options: Dict[str, Any]) -> None:
        super()._build(model_name, options)

        # pylint: disable=attribute-defined-outside-init
        self.tokenizer = AutoTokenizer.from_pretrained(
            model_name, use_fast=True, add_special_tokens=False
        )
        # Add the pad token to GPT2 dictionary.
        self.tokenizer.add_special_tokens({'additional_special_tokens': ["<|pad|>"]})
        self.tokenizer.pad_token = "<|pad|>"

        config = AutoConfig.from_pretrained(model_name)
        self.model = GPT2LMHeadModel.from_pretrained(model_name)
        # We need to resize the embedding layer because we added the pad token.
        self.model.resize_token_embeddings(len(self.tokenizer))
        self.model.eval()
        if "device" in options:
            self.model.to(options["device"])

    def _add_special_tokens(self, text: str) -> str:
        return self.tokenizer.bos_token + text + self.tokenizer.eos_token

    # @overrides
    def _tokens_log_prob_for_batch(
        self, text: List[str]
    ) -> List[Tuple[torch.FloatTensor, torch.LongTensor, List[str]]]:
        outputs:List[Tuple[torch.FloatTensor, torch.LongTensor, List[str]]] = []
        if len(text) == 0:
            return outputs

        # TODO: Handle overflowing elements for long sentences
        text = list(map(self._add_special_tokens, text))
        encoding: BatchEncoding = self.tokenizer.batch_encode_plus(
            text, return_tensors="pt",
        )
        with torch.no_grad():
            ids = encoding["input_ids"].to(self.model.device)
            nopad_mask = ids != self.tokenizer.pad_token_id
            logits: torch.Tensor = self.model(ids)[0]

        for sent_index in range(len(text)):
            sent_nopad_mask = nopad_mask[sent_index]
            # len(tokens) = len(text[sent_index]) + 1
            sent_tokens = encoding.tokens(sent_index)[1:]
            # sent_ids.shape = [len(text[sent_index]) + 1]
            sent_ids = ids[sent_index, sent_nopad_mask][1:]
            # logits.shape = [len(text[sent_index]) + 1, vocab_size]
            sent_logits = logits[sent_index, sent_nopad_mask][:-1, :]

            # ids_scores.shape = [seq_len + 1]
            sent_ids_scores = sent_logits.gather(1, sent_ids.unsqueeze(1)).squeeze(1)
            # log_prob.shape = [seq_len + 1]
            sent_log_probs = sent_ids_scores - sent_logits.logsumexp(1)

            sent_log_probs = cast(torch.FloatTensor, sent_log_probs)
            sent_ids = cast(torch.LongTensor, sent_ids)

            output = (sent_log_probs, sent_ids, sent_tokens)
            outputs.append(output)

        return outputs

    # @overrides
    @classmethod
    def _supported_model_names(cls) -> Iterable[str]:
        return GPT2LMHeadModel.pretrained_model_archive_map.keys()

```

**Comment by dldk-gael** [comment]
Hi Simone, 

Ok, I will update the PR to match the new API today. 
I think I found the problem in the code you proposed: you have to remove the score associated with the padding_token_id (in the vocabulary dim) before the logsumexp. I obtain the right score when doing that. So I will simply use your code with the correction in the PR

**Comment by simonepri** [comment]
> you have to remove the score associated with the padding_token_id (in the vocabulary dim) before the logsumexp.

Right! Thanks for the catch!

> So I will simply use your code with the correction in the PR
Perfect!

**Comment by simonepri** [comment]
Can you can force push the `upstream/master` into this branch? (make sure to backup the changes).
Your branch has too many commits.

If you prefere to keep your branch, you can create a new branch from the current master, add the changes again and open a new PR instead.

Also the additional tests you add are not necessary, you can remove them.
They are already implemented [here](https://github.com/simonepri/lm-scorer/blob/master/tests/models/abc/test_batch.py).

**Comment by dldk-gael** [comment]
Ok I go for second option.
